File: packages/python/spaces/starfish_spaces/node_access_revoked_error.py
"""NodeAccessRevokedError — raised when the server returns 403 on a node keyring pull."""

from __future__ import annotations


class NodeAccessRevokedError(Exception):
    """Raised when the server returns 403 on a node keyring fetch, indicating
    the requester's access to this node was revoked (e.g. the desk archived the
    ticket).

    This is a standalone class, NOT a subclass of :class:`SpaceAccessError`,
    so existing ``except SpaceAccessError`` soft-paths do not swallow it.

    Attributes:
        space_id: The space containing the revoked node.
        node_id:  The node whose access was revoked.
    """

    def __init__(self, space_id: str, node_id: str) -> None:
        self.space_id = space_id
        self.node_id = node_id
        super().__init__(f"Node {node_id!r} access revoked in space {space_id!r}")
        self.name = "NodeAccessRevokedError"


__all__ = ["NodeAccessRevokedError"]

File: packages/ts/spaces/src/node-access-revoked-error.ts
/**
 * `NodeAccessRevokedError` — thrown when the server returns 403 on a node
 * keyring pull, indicating the requester's access was revoked (e.g. the desk
 * archived the ticket). Callers should treat this as a terminal signal and
 * stop polling rather than retrying.
 *
 * IMPORTANT: this is a standalone class, NOT a subclass of `SpaceAccessError`.
 * This prevents existing `catch (SpaceAccessError)` soft-paths from silently
 * swallowing it.
 */
export class NodeAccessRevokedError extends Error {
  readonly spaceId: string
  readonly nodeId: string

  constructor(spaceId: string, nodeId: string) {
    super(`Node ${nodeId} access revoked in space ${spaceId}`)
    this.name = "NodeAccessRevokedError"
    this.spaceId = spaceId
    this.nodeId = nodeId
    Object.setPrototypeOf(this, NodeAccessRevokedError.prototype)
  }
}
